ompi/mca/coll/basic/coll_basic_gatherv.c

```diff
@@ -42,75 +42,102 @@ mca_coll_basic_gatherv_intra(const void *sbuf, int scount,
-                                     MCA_PML_BASE_SEND_STANDARD, comm));
+        err = ompi_datatype_get_extent(rdtype, &lb, &extent);
+        if (OMPI_SUCCESS != err) {
+            return OMPI_ERROR;
         }
-        return MPI_SUCCESS;
-    }
 
-    /* I am the root, loop receiving data. */
+        if (MPI_IN_PLACE != sbuf && (0 < scount) && (0 < rcounts[rank])) {
+            /* Directly copy self sbuf to rbuf */
+            err = ompi_datatype_sndrcv(sbuf, scount, sdtype,
+                                       ((char *) rbuf) + (extent * disps[rank]), rcounts[rank],
+                                       rdtype);
+            if (MPI_SUCCESS != err) {
+                return err;
+            }
+        }
 
-    ompi_datatype_type_size(rdtype, &rdsize);
-    if (OPAL_UNLIKELY(0 == rdsize)) {
-        /* bozzo case */
-        return MPI_SUCCESS;
-    }
+        ompi_request_t **reqs;
+        size_t nrecv = 0, recv_iter = 0;
 
-    err = ompi_datatype_get_extent(rdtype, &lb, &extent);
-    if (OMPI_SUCCESS != err) {
-        return OMPI_ERROR;
-    }
+        for (i = 0; i < size; ++i) {
+            /* We directly copied the data from self */
+            if (0 < rcounts[i] && rank != i) {
+                ++nrecv;
+            }
+        }
 
-    for (i = 0; i < size; ++i) {
-        ptmp = ((char *) rbuf) + (extent * disps[i]);
+        if (0 == nrecv) {
+            /* Nothing to receive */
+            return MPI_SUCCESS;
+        }
 
-        if (i == rank) {
-            /* simple optimization */
-            if (MPI_IN_PLACE != sbuf && (0 < scount) && (0 < rcounts[i])) {
-                err = ompi_datatype_sndrcv(sbuf, scount, sdtype,
-                                      ptmp, rcounts[i], rdtype);
-            }
-        } else {
+        reqs = ompi_coll_base_comm_get_reqs(module->base_data, nrecv);
+
+        for (i = 1; i < size; ++i) {
+            peer = (rank + i) % size;
+            ptmp = ((char *) rbuf) + (extent * disps[peer]);
             /* Only receive if there is something to receive */
-            if (rcounts[i] > 0) {
-                err = MCA_PML_CALL(recv(ptmp, rcounts[i], rdtype, i,
-                                        MCA_COLL_BASE_TAG_GATHERV,
-                                        comm, MPI_STATUS_IGNORE));
+            if (0 < rcounts[peer]) {
+                err = MCA_PML_CALL(irecv(ptmp, rcounts[peer], rdtype, peer,
+                                         MCA_COLL_BASE_TAG_GATHERV, comm, &reqs[recv_iter++]));
             }
         }
 
-        if (MPI_SUCCESS != err) {
-            return err;
+        assert(nrecv == recv_iter);
+
+        err = ompi_request_wait_all(nrecv, reqs, MPI_STATUSES_IGNORE);
+
+        if (MPI_ERR_IN_STATUS == err) {
+            for (int i = 0; i < nrecv; i++) {
+                if (MPI_REQUEST_NULL == reqs[i])
+                    continue;
+                if (MPI_ERR_PENDING == reqs[i]->req_status.MPI_ERROR)
+                    continue;
```

Comment:
Is this the right way to handle this case? Seems like we'd be leaking the request here...

Reply:
I read this comment
```
                     * Per MPI 2.2 p 60:
                     * Allows requests to be marked as MPI_ERR_PENDING if they are
                     * "neither failed nor completed." Which can only happen if
                     * there was an error in one of the other requests.
                     */
```

So if we encounter `MPI_ERR_PENDING` there should be an error somewhere else, so we exit on L117.

The requests are eventually freed on L122.